### ISSUE TITLE
Image conventions

### ISSUE THREAD

**Issue by Opioid** [ideasman42/png-encode-mini-rs#1]
I like this code because it keeps the dependencies low!

As I don't know anything about the rust ecosystem I have 2 questions;

1. Is it to be expected that RGBA pixels are stored as u32 instead of [u8, u8, u8, u8]? What is the rationale for write_rgba_from_u32()?
2. Are images in other rust libs usually stored in such a way, that rows stored at a higher memory adress should appear at the beginning of the image? Why are the rows exported in reverse order?

**Comment by ideasman42** [comment]
> I like this code because it keeps the dependencies low!
> 
> As I don't know anything about the rust ecosystem I have 2 questions;
> 
>     1. Is it to be expected that RGBA pixels are stored as u32 instead of [u8, u8, u8, u8]? What is the rationale for write_rgba_from_u32()?

Yes, otherwise use: `write_rgba_from_u8`, it's just for convenience to provide both.

>     2. Are images in other rust libs usually stored in such a way, that rows stored at a higher memory adress should appear at the beginning of the image? Why are the rows exported in reverse order?

This depends, I doubt there is a convention regarding Rust specifically, it's not uncommon to need to flip an image on the Y axis depending on the format and usage.

**Comment by Opioid** [comment]
> Yes, otherwise use: `write_rgba_from_u8`, it's just for convenience to provide both.
 
I will use write_rgba_from_u8(). I believe sticking to u8 would make it easier to provide a generic API that handles grayscale, rgb, rgba.

> This depends, I doubt there is a convention regarding Rust specifically, it's not uncommon to need to flip an image on the Y axis depending on the format and usage.

I have dealt with this often. I was just surprised to see that the rows are reversed by default, even though  apparently there is no specific reason for this. Perhaps this code was spun out of a project that stored images like this?
